File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/array/SubArrayTest.java
package com.fincity.nocode.kirun.engine.function.system.array;

import static org.junit.jupiter.api.Assertions.*;

import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.exception.KIRuntimeException;
import com.fincity.nocode.kirun.engine.runtime.FunctionExecutionParameters;
import com.google.gson.JsonArray;
import com.google.gson.JsonPrimitive;

class SubArrayTest {

	@Test
	void test() {

		JsonArray arr = new JsonArray();
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('d');
		arr.add('l');
		arr.add('d');
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('e');
		arr.add('d');

		JsonArray res = new JsonArray();

		res.add('b');
		res.add('c');
		res.add('d');
		res.add('l');
		res.add('d');
		res.add('a');
		res.add('b');

		FunctionExecutionParameters fep = new FunctionExecutionParameters()
				.setArguments(Map.of("source", arr, "findFrom", new JsonPrimitive(1), "length", new JsonPrimitive(7)));

		SubArray sub = new SubArray();

		assertEquals(
				sub.execute(fep).allResults().get(0).getResult().get( "output"), res);
	}

	@Test
	void test2() {

		JsonArray arr = new JsonArray();
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('d');
		arr.add('l');
		arr.add('d');
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('e');
		arr.add('d');

		JsonArray res = new JsonArray();

		res.add('a');
		res.add('b');
		res.add('c');
		res.add('d');
		res.add('l');
		res.add('d');
		res.add('a');
		res.add('b');
		res.add('c');
		res.add('e');
		res.add('d');

		FunctionExecutionParameters fep = new FunctionExecutionParameters().setArguments(
				Map.of("source", arr, "findFrom", new JsonPrimitive(1123), "length", new JsonPrimitive(arr.size())));

		SubArray sub = new SubArray();

		assertThrows(KIRuntimeException.class, () -> sub.execute(fep).allResults().get(0).getResult().get("output"));
	}

	@Test
	void test3() {

		JsonArray arr = new JsonArray();
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('d');
		arr.add('l');
		arr.add('d');
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('e');
		arr.add('d');

		JsonArray res = new JsonArray();

		res.add('l');
		res.add('d');
		res.add('a');
		res.add('b');

		FunctionExecutionParameters fep = new FunctionExecutionParameters()
				.setArguments(Map.of("source", arr, "findFrom", new JsonPrimitive(4), "length", new JsonPrimitive(4)));

		SubArray sub = new SubArray();

		sub.execute(fep);

		assertEquals(sub.execute(fep).allResults().get(0).getResult().get( "output"), res);
	}

	@Test
	void test4() {
		JsonArray arr = new JsonArray();
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('d');
		arr.add('l');
		arr.add('d');
		arr.add('a');
		arr.add('b');
		arr.add('c');
		arr.add('e');
		arr.add('d');

		JsonArray res = new JsonArray();

		res.add('l');
		res.add('d');
		res.add('a');
		res.add('b');

		FunctionExecutionParameters fep = new FunctionExecutionParameters()
				.setArguments(Map.of("source", arr, "findFrom", new JsonPrimitive(4), "length", new JsonPrimitive(4)));

		SubArray sub = new SubArray();

		

		assertEquals(sub.execute(fep).allResults().get(0).getResult().get("output"), res);

	}

	@Test
	void test5() {

		var array1 = new JsonArray();

		array1.add("test");
		array1.add("Driven");
		array1.add("developement");
		array1.add("I");
		array1.add("am");
		array1.add("using");
		array1.add("eclipse");
		array1.add("I");
		array1.add("to");
		array1.add("test");
		array1.add("the");
		array1.add("changes");
		array1.add("with");
		array1.add("test");
		array1.add("Driven");
		array1.add("developement");

		var array2 = new JsonArray();

		array2.add("test");
		array2.add("Driven");
		array2.add("developement");
		array2.add("I");
		array2.add("am");
		array2.add("using");
		array2.add("eclipse");
		array2.add("I");
		array2.add("to");
		array2.add("test");
		array2.add("the");
		array2.add("changes");
		array2.add("with");

		var array3 = new JsonArray();

		array3.add("test");
		array3.add("Driven");
		array3.add("developement");
		array3.add("I");
		array3.add("am");
		array3.add("using");
		array3.add("eclipse");
		array3.add("I");
		array3.add("to");
		array3.add("test");
		array3.add("the");
		array3.add("changes");
		array3.add("with");
		array3.add("test");
		array3.add("Driven");
		array3.add("developement");

		var array4 = new JsonArray();

		array4.add("test");
		array4.add("Driven");
		array4.add("developement");
		array4.add("I");
		array4.add("am");
		array4.add("using");
		array4.add("eclipse");
		array4.add("I");
		array4.add("to");

		var arr = new JsonArray();
		arr.add(array2);
		arr.add(array4);
		arr.add(array1);
		arr.add(array1);
		arr.add(array1);
		arr.add(array3);
		arr.add(array2);
		arr.add(array4);
		arr.add(array1);
		arr.add(array1);
		arr.add(array4);

		var res = new JsonArray();
		res.add(array1);
		res.add(array1);
		res.add(array1);

		FunctionExecutionParameters fep = new FunctionExecutionParameters()
				.setArguments(Map.of("source", arr, "findFrom", new JsonPrimitive(2), "length", new JsonPrimitive(3))) ;

		SubArray mis = new SubArray();


		assertEquals(mis.execute(fep).allResults().get(0).getResult().get("output"), res);

	}
}

File: kirun-js/__tests__/engine/function/system/array/SubArrayTest.ts
import { SubArray } from '../../../../../src/engine/function/system/array/SubArray';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';

test('SubArray of Test 1', () => {
    let sub: SubArray = new SubArray();

    let array: any[] = [];

    array.push('test');
    array.push('Driven');
    array.push('developement');
    array.push('I');
    array.push('am');
    array.push('using');
    array.push('eclipse');
    array.push('I');
    array.push('to');
    array.push('test');
    array.push('the');
    array.push('changes');
    array.push('with');
    array.push('test');
    array.push('Driven');
    array.push('developement');

    let res: any[] = [];

    res.push('am');
    res.push('using');
    res.push('eclipse');
    res.push('I');
    res.push('to');
    res.push('test');
    res.push('the');
    res.push('changes');
    res.push('with');

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(
            new Map<string, any>([
                [SubArray.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
                [SubArray.PARAMETER_INT_FIND_FROM.getParameterName(), 4],
                [SubArray.PARAMETER_INT_LENGTH.getParameterName(), 9],
            ]),
        )
        .setOutput(new Map([]))
        .setContext(new Map([]));

    sub.execute(fep);

    expect(array).toStrictEqual(res);
});

test('SubArray of Test 2', () => {
    let sub: SubArray = new SubArray();

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(
            new Map<string, any>([
                [SubArray.PARAMETER_ARRAY_SOURCE.getParameterName(), undefined],
                [SubArray.PARAMETER_INT_FIND_FROM.getParameterName(), 4],
                [SubArray.PARAMETER_INT_LENGTH.getParameterName(), 9],
            ]),
        )
        .setOutput(new Map([]))
        .setContext(new Map([]));

    expect(() => sub.execute(fep)).toThrow;
});

test('SubArray of Test 2', () => {
    let sub: SubArray = new SubArray();

    let array: any[] = [];
    array.push('a');
    array.push('b');
    array.push('c');
    array.push('d');
    array.push('l');
    array.push('d');
    array.push('a');
    array.push('b');
    array.push('c');
    array.push('e');
    array.push('d');

    let res: any[] = [];
    res.push('b');
    res.push('c');
    res.push('d');
    res.push('l');
    res.push('d');
    res.push('a');
    res.push('b');

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(
            new Map<string, any>([
                [SubArray.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
                [SubArray.PARAMETER_INT_FIND_FROM.getParameterName(), 1],
                [SubArray.PARAMETER_INT_LENGTH.getParameterName(), 7],
            ]),
        )
        .setOutput(new Map([]))
        .setContext(new Map([]));

    sub.execute(fep);

    expect(array).toStrictEqual(res);
});

test('SubArray of Test 3', () => {
    let sub: SubArray = new SubArray();

    let array: any[] = [];
    array.push('a');
    array.push('b');
    array.push('c');
    array.push('d');
    array.push('l');
    array.push('d');
    array.push('a');
    array.push('b');
    array.push('c');
    array.push('e');
    array.push('d');

    let res: any[] = [];
    res.push('b');
    res.push('c');
    res.push('d');
    res.push('l');
    res.push('d');
    res.push('a');
    res.push('b');

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(
            new Map<string, any>([
                [SubArray.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
                [SubArray.PARAMETER_INT_FIND_FROM.getParameterName(), 1123],
                [SubArray.PARAMETER_INT_LENGTH.getParameterName(), 7],
            ]),
        )
        .setOutput(new Map([]))
        .setContext(new Map([]));

    expect(() => sub.execute(fep)).toThrow;
});

test('SubArray of Test 4', () => {
    let sub: SubArray = new SubArray();

    let array1: any[] = [];
    array1.push('test');
    array1.push('Driven');
    array1.push('developement');
    array1.push('I');
    array1.push('am');
    array1.push('using');
    array1.push('eclipse');
    array1.push('I');
    array1.push('to');
    array1.push('test');
    array1.push('the');
    array1.push('changes');
    array1.push('with');
    array1.push('test');
    array1.push('Driven');
    array1.push('developement');

    let array2: any[] = [];
    array2.push('test');
    array2.push('Driven');
    array2.push('developement');
    array2.push('I');
    array2.push('am');
    array2.push('using');
    array2.push('eclipse');
    array2.push('I');
    array2.push('to');
    array2.push('test');
    array2.push('the');
    array2.push('changes');
    array2.push('with');

    let array3: any[] = [];

    array3.push('test');
    array3.push('Driven');
    array3.push('developement');
    array3.push('I');
    array3.push('am');
    array3.push('using');
    array3.push('eclipse');
    array3.push('I');
    array3.push('to');
    array3.push('test');
    array3.push('the');
    array3.push('changes');
    array3.push('with');
    array3.push('test');
    array3.push('Driven');
    array3.push('developement');

    let array4: any[] = [];
    array4.push('test');
    array4.push('Driven');
    array4.push('developement');
    array4.push('I');
    array4.push('am');
    array4.push('using');
    array4.push('eclipse');
    array4.push('I');
    array4.push('to');

    let array: any[] = [];
    array.push(array2);
    array.push(array4);
    array.push(array1);
    array.push(array1);
    array.push(array1);
    array.push(array3);
    array.push(array2);
    array.push(array4);
    array.push(array1);
    array.push(array1);
    array.push(array4);

    let res: any[] = [];
    res.push(array2);
    res.push(array4);
    res.push(array1);

    let fep: FunctionExecutionParameters = new FunctionExecutionParameters()
        .setArguments(
            new Map<string, any>([
                [SubArray.PARAMETER_ARRAY_SOURCE.getParameterName(), array],
                // [SubArray.PARAMETER_INT_FIND_FROM.getParameterName(), 2],
                [SubArray.PARAMETER_INT_LENGTH.getParameterName(), 3],
            ]),
        )
        .setOutput(new Map([]))
        .setContext(new Map([]));

    sub.execute(fep);

    expect(array).toStrictEqual(res);
});
